sky/provision/azure/instance.py

```diff
@@ -714,62 +717,83 @@ def open_ports(
     subscription_id = provider_config['subscription_id']
     resource_group = provider_config['resource_group']
     network_client = azure.get_client('network', subscription_id)
+    nsg_name = _NSG_NAME.format(cluster_name_on_cloud=cluster_name_on_cloud)
```

Comment:
@Michaelvll Thanks! I have two questions to address:

1. If I understand correctly, it doesn't seem like we can keep the original order of `while` and `for` loop since we need to call `list_network_security_groups(resource_group)` within the `while` loop(`while not nsg_created:`) so we can update the iterator from the method when the status check, `if nsg.provisioning_state not in ['Creating', 'Updating']:`, fails within the for-loop on the iterator(`for nsg in nsg_list:`). But, I may be missing something. Please help to point out if so.

2. I can remove the check for specific nsg name, but this was also intended to be used for #3764 as well. As you mentioned, we currently use one resource group per cluster, but this changes when resource group specification is used with `sky serve`. Serve controller and all the replicas would be created under the user specified resource group, and in this case, there would be multiple NSGs under the resource group. This was causing issues as opening ports while launching replica 4 was checking on NSG creation for replica 5 and got stuck. 

    If losing track of the NSG name for the cluster throughout the `launch` cycle is your concern, we can keep on track of the NSG name while passing the name as a parameter to our ARM template instead of directly creating it from the ARM(`"nsgName": "[concat('sky-', parameters('clusterId'), '-nsg')]"`) and perhaps have that stored in `provider_config` as well, so we can read from it instead of using `_NSG_NAME.format()`.

    2-1. Is it possible that we may need multiple NSGs for a single cluster in near future? We don't seem to do this now.

Reply:
For 1, can we check if there is an Azure API that get a single `nsg` instead of listing all of them, so that we can keep getting that nsg within the while loop? The current loop order is unintuitive, looping through the nsg list within a retry loop. It would be more readable to do the retry loop for each NSG (although we only select a single NSG).

For 2, it makes sense when a resource group can be specified. In that case, how about we just create a function in `azure/config.py` that generates the NSG name, so it can be reused in both `config.py` and here? We have to be careful with backward compatibility then, if we do the exact match of the NSG. How about we allow a list of NSG names to support legacy names, such as `ray-{clusterId}-...` and the one that will be added with #3848

Reply:
@Michaelvll Following your review, I found an API method `.get` to directly retrieve the nsg object. And there's an issue of using this with multiple possible NSG names(legacy names). Attempting to get an nsg instance with the following snippet takes about ~15s to return a `ResourceNotFoundError` when given a non-existent nsg name:
```
nsg = network_client.network_security_groups.get(resource_group_name, nsg_name)
```
And we have to provide non-existent nsg name at most twice with two legacy nsg names, which adds ~30s, to provision VM. So, I instead kept the listing method to iterate through the NSG list to find a matching name among our legacy and current nsg names. This is done in a separate method, `_get_cluster_nsg`, so it also resolves your concern on readability.

Also, following your comment, I created a method to create NSG name with our current naming convention and used it within `config.py::bootstrap_instances` and `instance.py::_get_cluster_nsg`, which is called by `instance.py::open_ports`.

These are done at [befcbc1](https://github.com/skypilot-org/skypilot/pull/3844/commits/befcbc1db99c1feb5773638e9c8cd53d12a57ccf)